feat: スレッドのメッセージ表示失敗時のエラー表示処理を実装

harumoren2507 opened the pull request:
# エラーハンドリング機能の実装

## 概要
スレッド画面でのメッセージ取得失敗時に発生していたアプリクラッシュとエラー通知不備を解決しました。

## 課題
- メッセージ取得失敗時にアプリがクラッシュ
- エラー内容がユーザーに伝わらない
- 通信エラーなどによるクラッシュがユーザー体験上良くない。

## 実装内容

### 1. データ取得処理の安全化
- `Result`型ベースのエラーハンドリングを導入
- 通信失敗時も安全に結果を返却する仕組みを構築

### 2. ユーザー向けエラー通知

[要件]
- ViewModelでエラー検知機能を実装
- Snackbarによる「メッセージを取得できませんでした」の表示

[要件ではないが工夫したこと]
- エラー原因を日本語を表示を行い、ユーザビリティを向上。
- ユーザーはメッセージ取得できなかったことだけでなく、原因も知りたいだろうと考えた。

## 技術的選択の理由

### Snackbar採用理由
要件ヒント「ユーザーの操作を妨げないこと」を重視し、以下の理由でSnackbarを選択：
- 数秒後の自動消失
- Dialogと比較してユーザー操作の継続性を保持
- 最初に読んだ記事を見て、https://developer.android.com/develop/ui/compose/components/snackbar?hl=ja
snackbarを使いましたが、今回のようなエラー表示のような簡潔なものならtoast でも良かった気がしてきました。再試行ボタンとか今回入れる余地はなさそうなので。


## 動作検証手順
1. kintone spaceをスレッド表示可能な通信環境で起動
2. アプリを落とさずに、上部バーなどでWi-Fi/5G接続を無効化
3. 任意のスレッドを選択
4. 「メッセージを表示できませんでした」の表示を確認
<img width="273" height="575" alt="Screenshot 2025-08-19 at 12 28 00" src="https://github.com/user-attachments/assets/bf2aff67-fe5e-498c-8b0a-722769af36c2" />


## 期待される効果
- アプリクラッシュの根絶
- エラー発生時のユーザー体験向上

harumoren2507 committed:
feat: スレッド画面のメッセージ取得エラーを改善

これまで、スレッド画面でメッセージ取得に失敗すると、アプリがクラッシュしたり、エラーがユーザーに伝わらなかったりしていました。

この問題を解決するため、エラーハンドリングを全面的に見直し、ユーザーに分かりやすい形でエラーを通知するようにしました。

- データ取得処理は`Result`型ベースに刷新され、通信失敗時も安全に結果を返せるようになりました。
- ViewModelはエラーを検知し、ユーザーに「メッセージを取得できませんでした」とSnackbarで表示するようになりました。
- エラー原因も日本語で簡潔に表示されることで、ユーザーが状況を把握しやすくなると考えます。
- 画面のリコンポジションでSnackbarが重複表示されないよう、`LaunchedEffect`などで制御されるようになりました。

これにより、アプリの安定性とユーザー体験が向上すると考えます。

harumoren2507 committed:
fix:lintにてフォーマットを行った

harumoren2507 committed:
テストリポジトリの戻り値型を修正

- SpaceViewModelTestのFakeRepositoryの戻り値型を更新
- ThreadViewModelTestのFakeRepositoryの戻り値型を更新
- List<ThreadMessage>からResult<List<ThreadMessage>>に変更
- インターフェースとの不一致によるコンパイルエラーを解消

shinsuke-fujita1126 commented on data/space/src/main/java/com/cybozu/sample/kintone/spaces/data/space/SpaceRepositoryImpl.kt:
```diff
@@ -9,6 +9,8 @@ internal class SpaceRepositoryImpl @Inject constructor(
-    override suspend fun getMessagesForThread(threadId: String): List<ThreadMessage> =
-        spaceRemoteDataSource.getMessagesForThread(threadId = threadId).result.items
+    override suspend fun getMessagesForThread(threadId: String): Result<List<ThreadMessage>> =
+        runCatching {
```
runCatchingを使うアイデアいいですね。
ただ、コルーチンがキャンセルされる時にCancellationExceptionがスローされるのですが、こいつはrethrowしないとダメなのですが、runCatchingだとrethrowされないのでキャンセルが正しくされないです。
https://developer.android.com/kotlin/coroutines/coroutines-best-practices?hl=ja#exceptions

nowinandroidという有名なリポジトリではsuspendRunCatchingというCancellationExceptionを考慮したrunCatchingがあるので、同じようなものを用意して使うと良いですね。
https://github.com/android/nowinandroid/blob/main/core/data/src/main/kotlin/com/google/samples/apps/nowinandroid/core/data/SyncUtilities.kt#L55

shinsuke-fujita1126 commented on feature/communicate/src/main/java/com/cybozu/sample/kintone/spaces/feature/communicate/thread/ThreadScreen.kt:
```diff
@@ -59,7 +63,8 @@ fun ThreadScreen(
         threadName = threadName,
-        uiState = uiState
+        uiState = uiState,
+        onErrorMessageShown = viewModel::errorMessageShown
```
[good]
エラーメッセージが表示された後のuiStateの更新も意識できてるのがいいですね。

shinsuke-fujita1126 commented on feature/communicate/src/main/java/com/cybozu/sample/kintone/spaces/feature/communicate/thread/ThreadViewModel.kt:
```diff
@@ -29,15 +29,37 @@ class ThreadViewModel @AssistedInject constructor(
         }
     }
+
+    private fun getLocalizedErrorMessage(throwable: Throwable): String =
```
[good]
エラーの種類を意識したエラーメッセージの出しわけいいですね！

shinsuke-fujita1126 commented on feature/communicate/src/main/java/com/cybozu/sample/kintone/spaces/feature/communicate/thread/ThreadViewModel.kt:
```diff
@@ -29,15 +29,37 @@ class ThreadViewModel @AssistedInject constructor(
+        when (throwable) {
+            is java.net.UnknownHostException -> "サーバーが見つかりません"
+            is java.net.SocketTimeoutException -> "通信がタイムアウトしました"
+            is java.io.IOException -> "ネットワークエラーが発生しました"
```
[任意]
好みによるかもですが、以下のように書くこともできます
```suggestion
            is UnknownHostException -> "サーバーが見つかりません"
            is SocketTimeoutException -> "通信がタイムアウトしました"
            is IOException -> "ネットワークエラーが発生しました"
```

harumoren2507 committed:
CancellationExceptionの問題を修正

- suspendRunCatchingを追加してCancellationExceptionを再スロー
- SpaceRepositoryImplで使用してコルーチンキャンセルを正常化

harumoren2507 committed:
fix:lintでの修正

harumoren2507 committed:
例外クラスのimportを明示化

- java.net.*, java.io.*の完全修飾名を短縮名に変更
- 必要なimport文を追加

shinsuke-fujita1126 reviewed (approved):
LGTM!
